### PR TITLE
Prototype pattern added

### PR DESCRIPTION
**Pull request by Makas9**
It may be not practical, but we need to meet the deadlines. Added few functions to BaseShape just because of requirement to compare between Shallow copies (Prototype pattern) and Deep copies. Added LabTest class for demo during lab defense.

### CODE REVIEW

**Comment by Makas9** [comment]
I'll leave GetHashCode() for now

File: Core/NoOpRunner.Core/Shapes/BaseShape.cs [hunk 1]
```diff
@@ -74,5 +74,21 @@ public virtual void OnClick()
...
+
+        public BaseShape Clone()
+        {
+            return (BaseShape)this.MemberwiseClone();
+        }
+        public BaseShape DeepCopy()
+        {
+            BaseShape clone = (BaseShape)this.MemberwiseClone();
+            clone.ShapeBlocks = new List<ShapeBlock>(ShapeBlocks);
+            return clone;
```
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/BaseShape.cs]: This is not a deep copy. The copy will share the same references to ShapeBlocks. You can check this by creating a copy of a BaseShape, and then modifying an offset of a shapeBlock from the original shape. The "deep" copy will also reflect the changes: 
```
BaseShape deep = shape.DeepCopy();
var sp = shape.GetShapes().First();
var sp1 = deep.GetShapes().First();
sp.OffsetX = 55;
var trueDeep = sp1.OffsetX != sp.OffsetX;```

**Review by Kodnot — CHANGES_REQUESTED** [review]
Not convinced you can defend this impl, but that's not my problem. LabTest.cs code isn't really part of the core database, so I'll refrain from nitpicking on it.

File: Core/NoOpRunner.Core/Shapes/BaseShape.cs (hunk 1 above)
Comment (Makas9) [reply Core/NoOpRunner.Core/Shapes/BaseShape.cs]: Fixed

File: Core/NoOpRunner.Core/Shapes/BaseShape.cs [hunk 2]
```diff
@@ -74,5 +74,29 @@ public virtual void OnClick()
...
+        {
+            BaseShape clone = (BaseShape)this.MemberwiseClone();
+            clone.ShapeBlocks = new List<ShapeBlock>();
+            foreach (ShapeBlock shape in ShapeBlocks)
+            {
+                ShapeBlock block = new ShapeBlock();
+                block.OffsetX = shape.OffsetX;
+                block.OffsetY = shape.OffsetY;
+                clone.ShapeBlocks.Add(block);
+            }
```
Comment (Kodnot) [inline Core/NoOpRunner.Core/Shapes/BaseShape.cs]: Could just be `clone.ShapeBlocks = ShapeBlocks.Select(x => new ShapeBlock() { OffsetX = x.OffsetX, OffsetY = x.OffsetY }).ToList();`

**Review by Kodnot — APPROVED** [review]
As long as you're fine with it

**Pull request merged**